app/vmalert/config/config.go

```diff
@@ -298,16 +299,34 @@ func parseConfig(data []byte) ([]Group, error) {
 	if err != nil {
 		return nil, fmt.Errorf("cannot expand environment vars: %w", err)
 	}
-	g := struct {
-		Groups []Group `yaml:"groups"`
-		// Catches all undefined fields and must be empty after parsing.
-		XXX map[string]any `yaml:",inline"`
-	}{}
-	err = yaml.Unmarshal(data, &g)
-	if err != nil {
-		return nil, err
+
+	var allGroups []Group
+
+	decoder := yaml.NewDecoder(bytes.NewReader(data))
+	for {
+		var g struct {
+			Groups []Group        `yaml:"groups"`
+			XXX    map[string]any `yaml:",inline"` // Catches all undefined fields and must be empty after parsing.
+		}
+
+		// Decode the next document
+		if err := decoder.Decode(&g); err != nil {
+			if err == io.EOF { // End of file indicates no more documents to read
+				break
+			}
+			return nil, fmt.Errorf("failed to decode YAML document: %w", err)
+		}
+
+		// Check for unknown fields
+		if err := checkOverflow(g.XXX, "config"); err != nil {
+			return nil, err
+		}
+
+		// Append valid groups
+		allGroups = append(allGroups, g.Groups...)
```

Comment:
WDYT the following changes (optional, just for concise):
1. move struct definition outside the `for` loop.
2. collect all unknown fields to `overflowMap`, instead of exiting the loop with partial unknown fields error returned. And we can keep the `return g.Groups, checkOverflow(g.XXX, "config")` at the end of this function unchanged.

```suggestion
	type config struct {
		Groups []Group        `yaml:"groups"`
		XXX    map[string]any `yaml:",inline"` // Catches all undefined fields and must be empty after parsing.
	}
	overflowMap := make(map[string]any)
	for {
		var g config

		// Decode the next document
		if err := decoder.Decode(&g); err != nil {
			if err == io.EOF { // End of file indicates no more documents to read
				break
			}
			return nil, fmt.Errorf("failed to decode YAML document: %w", err)
		}
		maps.Copy(overflowMap, g.XXX)

		// Append valid groups
		allGroups = append(allGroups, g.Groups...)
```

Reply:
Hi @jiekun, 

1.  move struct definition outside the for loop -> Sure done
2. collect all unknown fields to overflowMap .. -> Sure, I will skip the unknown fields and send `g.Groups` but why to store them in a map? Do we log them somewhere, as we will only be returning the correct parsed groups

Reply:
> but why to store them in a map?
  1. This function will return an error when invalid field exists. In this case, the `allGroups` won't be used by caller.
  2. I want to collect all invalid fields instead of breaking the loop and return invalid fields in current doc. So `overflowMap` is defined outside the for-loop and checked by `return g.Groups, checkOverflow(overflowMap, "config")` (sorry for the typo, it was wrongly suggested before. I just updated my suggested change, please see below.).